### PR TITLE
Overload `MIN/MAX/MAX_BY/MIN_BY` to return the "top" `N` values

### PR DESCRIPTION
**Pull request by Maxxen**
This adds an additional overload to the `MIN/MAX/MAX_BY/MIN_BY/ARG_MIN/ARG_MAX` aggregate functions that take an `N` integer as the last argument and returns the "top"/"bottom" `N` values as a list instead of a single value. 

Generally computing the "top N in each group" in SQL is a [very common](https://stackoverflow.com/search?q=sql+top+n+in+group&s=f0d67570-1567-4fec-b01b-16825782b205) but somewhat complicated operation, usually requiring a subquery + window operation to order and filter on row number, e.g.

## Example

```sql

-- Given the table
FROM t1;
----
┌─────────┬───────┐
│   grp   │  val  │
│ varchar │ int32 │
├─────────┼───────┤
│ a       │     2 │
│ a       │     1 │
│ b       │     5 │
│ b       │     4 │
│ a       │     3 │
│ b       │     6 │
└─────────┴───────┘

-- Get the top 3 'val's of each group 'grp'
SELECT rs.val, rs.grp 
FROM (
  SELECT val, grp, row_number() OVER (PARTITION BY grp ORDER BY val DESC) as rid
  FROM t1
) AS rs 
WHERE rid < 4;
-----
┌───────┬─────────┐
│  val  │   grp   │
│ int32 │ varchar │
├───────┼─────────┤
│     6 │ b       │
│     5 │ b       │
│     4 │ b       │
│     3 │ a       │
│     2 │ a       │
│     1 │ a       │
└───────┴─────────┘

-- Or even worse, if you want the values as lists

SELECT array_agg(rs.val), rs.grp 
FROM (
  SELECT val, grp, row_number() OVER (PARTITION BY grp ORDER BY val DESC) as rid 
  FROM t1 ORDER BY val DESC
) as rs 
WHERE rid < 4 
GROUP BY rs.grp;
----
┌───────────────────┬─────────┐
│ array_agg(rs.val) │   grp   │
│      int32[]      │ varchar │
├───────────────────┼─────────┤
│ [3, 2, 1]         │ a       │
│ [6, 5, 4]         │ b       │
└───────────────────┴─────────┘

-- Youch! 
-- But with this PR, you can simply do:

SELECT max(val, 3) FROM t1 GROUP BY gap;
----
┌─────────────┐
│ max(val, 3) │
│   int32[]   │
├─────────────┤
│ [3, 2, 1]   │
│ [6, 5, 4]   │
└─────────────┘

-- Also works with max_by/arg_max
SELECT arg_max(grp, val, 3) FROM t1;
----
┌──────────────────────┐
│ arg_max(grp, val, 3) │
│      varchar[]       │
├──────────────────────┤
│ [b, b, b]            │
└──────────────────────┘
```

## Implementation

The current implementation provides specializations for values (and every combination of value/argument pairs in case of `arg_max/min`) with a physical type of `INTEGER`, `BIGINT`, `DOUBLE`, `FLOAT`, `VARCHAR`, everything else falls back to a "sort key" serialization based implementation. We can extend this in the future and specialize for more primitive types.

The `VARCHAR` and fallback "sort key" specializations uses the aggregate arena allocator so all memory is tracked, and will also reuse the existing allocation in the priority queue heap slot whenever an entry is evicted and the new entry fits within the previous blob.

### CODE REVIEW

**Comment by carlopi** [comment]
Another comment from me: size with the changes in this PR bumps by +2%.

We don't have any CI that checks for this (but I guess we should), but in any case it's somewhat significant.

File: test/sql/aggregate/aggregates/arg_min_max_n_all_types.test_slow [hunk 1]
```diff
@@ -0,0 +1,49 @@
+# name: test/sql/aggregate/aggregates/arg_min_max_n_all_types.test_slow
+# description: Test approx top K operator
+# group: [aggregates]
+
+statement ok
```
[the remark this answers is not in the document]
Comment (Mytherin) [reply test/sql/aggregate/aggregates/arg_min_max_n_all_types.test_slow:5]: Perhaps we can use the query you wrote as verification in the tests, e.g.:

```sql
CREATE MACRO compute_top_k(table_name, group_col, val_col, k) AS TABLE
SELECT rs.grp, array_agg(rs.val)
FROM (
  SELECT group_col AS grp, val_col AS val, row_number() OVER (PARTITION BY group_col ORDER BY val_col DESC) as rid 
  FROM query_table(table_name::VARCHAR) ORDER BY group_col DESC
) as rs 
WHERE rid <= k
GROUP BY ALL
ORDER BY ALL;

SELECT * FROM compute_top_k(lineitem, l_returnflag, l_orderkey, 3);
┌─────────┬─────────────────────────────┐
│   grp   │      array_agg(rs.val)      │
│ varchar │           int64[]           │
├─────────┼─────────────────────────────┤
│ A       │ [5999975, 5999975, 5999968] │
│ N       │ [6000000, 6000000, 5999973] │
│ R       │ [5999975, 5999974, 5999974] │
└─────────┴─────────────────────────────┘
```

Should be identical to:

```sql
SELECT l_returnflag, max(l_orderkey, 3) FROM lineitem GROUP BY ALL ORDER BY ALL;
```

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks great - some comments from my side:

File: src/include/duckdb/core_functions/aggregate/minmax_n_helpers.hpp [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+			heap.back().second.Assign(allocator, value);
+			std::push_heap(heap.begin(), heap.end(), Compare);
+		}
+		D_ASSERT(std::is_heap(heap.begin(), heap.end(), Compare));
+	}
+
+	void Insert(ArenaAllocator &allocator, const BinaryAggregateHeap &other) {
+		for (auto &slot : other.heap) {
+			Insert(allocator, slot.first.value, slot.second.value);
+		}
```
[the remark this answers is not in the document]
Comment (carlopi) [reply src/include/duckdb/core_functions/aggregate/minmax_n_helpers.hpp:209]: Yep, we agreed with Max on this. I think benchmark (giving results are impressive), and indeed verification given two queries should be equivalent would be cool to have.

File: test/sql/aggregate/aggregates/arg_min_max_n_all_types.test_slow (hunk 1 above)
[the remark this answers is not in the document]
Comment (Maxxen) [reply test/sql/aggregate/aggregates/arg_min_max_n_all_types.test_slow:5]: Apologies, there were a couple of last minute commits I did while writing up the PR description but forgot to push before I opened this PR that fixes the correctness issue (I had the min/max heap ordering the wrong way around).

The `arg_min_max_n_all_types.test_slow` checks the equivalence between the manual window query and the new overloads in the ungrouped case for all types, but I'll see if I can add a test with grouping as well, perhaps, with the TPCH data.

**Comment by Maxxen** [comment]
Some micro benchmarks!

tpch SF=5

Using the window
```
benchmark/tpch/aggregate/top_n_in_group_window.benchmark        1       0.563334
benchmark/tpch/aggregate/top_n_in_group_window.benchmark        2       0.557412
benchmark/tpch/aggregate/top_n_in_group_window.benchmark        3       0.557143
benchmark/tpch/aggregate/top_n_in_group_window.benchmark        4       0.569500
benchmark/tpch/aggregate/top_n_in_group_window.benchmark        5       0.558921
```

Using the aggregate
```
benchmark/tpch/aggregate/top_n_in_group_agg.benchmark   1       0.193027
benchmark/tpch/aggregate/top_n_in_group_agg.benchmark   2       0.193580
benchmark/tpch/aggregate/top_n_in_group_agg.benchmark   3       0.192352
benchmark/tpch/aggregate/top_n_in_group_agg.benchmark   4       0.192490
benchmark/tpch/aggregate/top_n_in_group_agg.benchmark   5       0.192478
```

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**